PEP 605: Change ABI management to encourage pre-built wheels

ncoghlan opened the pull request:
The initial version of the PEP painted a pretty dire picture when it came
to the feasibility of supporting wheels built against the full CPython
ABI on the rolling beta release stream.

I wouldn't actually expect things to be that bad in reality, so this adjusts
the proposal to actively encourage the early creation of wheel archives
targeting the rolling releases.

zooba commented on pep-0605.rst:
```diff
@@ -24,32 +24,43 @@ pre-release management process as follows:
 
+Similarly, the guidance given for alpha releases would be amended to state
+"intended for library compatibility testing and the creation of ABI compatible
+binary artifacts", rather than simply saying "not for production use".
```
I think this point is important enough to put it right near the top as well.

zooba commented on pep-0605.rst:
```diff
@@ -184,11 +197,72 @@ The majority of the proposed changes in this PEP only affect the handling of
+binary compatibility against).
+
+Pre-freeze releases would gain an additional flag in their C ABI compatibility
+markers to avoid binary compatibility issues with the eventual stable release.
```
I still think this is unnecessary. The whole point of the beta builds is to be ABI compatible with the next release, so why would we mark them as incompatible?

Similarly, why would we mark alpha builds as incompatible with the compatible beta builds that follow them?

The best thing we could do here is add a version check into module initialisers with the full hex version so we can fail them quickly when a specific build has crossed an alpha release without being rebuilt. Anything else is just proving that we don't believe our own rhetoric about making compatible releases :)

ncoghlan replied:
The point is to make them fail at the point where we cross into X.Y.0rc1, not to make them fail on any of the alpha or beta releases - users of the beta releases are on their own in that regard.

If we were to put the "recompile the world, even if you don't actually need to" gate on every alpha release, that makes the rolling stream a PITA to support even for extension modules that aren't tinkering with struct internals, while if you don't have any technically enforced rebuild gate at all, then you risk extension modules built early in the X.Y pre-release cycle causing segfaults for people that are moving directly from the previous stable release to the X.Y.0 stable release.

ncoghlan replied:
I've added a note about this to the section on ABI management changes. It's certainly possible to leave it out, it just creates the risk of painful debugging session on the *stable* X.Y.0 release that get traced back to "Oh, our dependency 'superfast-binary-operation' was affected by a CPython ABI break in X.Y.0a5, but the project hasn't published a new build since then".

Whereas if the ABI tag changes, projects can publish a new wheel file post-rc1 without even needing to publish a new release of the project (since the ABI tags are different, the filenames are different, and hence can be added to an existing release), and if they haven't published one, users will fall back to building from source (as they do today), rather than downloading a potentially incompatible binary.

ncoghlan replied:
I've now also incorporated the example scenario from the Discourse thread, and made it clear that a smarter scheme that allowed the use of all binaries built since that last alpha release would also be fine (I just don't know how such a scheme could actually work, given that we don't know ahead of time which alpha release is going to be the last pre-release alpha)

zooba commented on pep-0605.rst:
```diff
@@ -460,48 +548,47 @@ interests of operating environments that are able to consume new releases as
-the fact that the release includes APIs and ABIs that aren't necessarily stable
-yet.
+only being put off by the "beta" label, then it may make sense to emphasise
+the "ABI breaking" and "binary compatible" names over the "alpha" and "beta"
```
👏 Nice

Maybe emphasise the initials of the two names to make the a/b clearer?

```suggestion
the "*A*BI breaking" and "*B*inary compatible" names over the "alpha" and "beta"
```

zooba commented on pep-0605.rst:
```diff
@@ -460,48 +548,47 @@ interests of operating environments that are able to consume new releases as
+"beta" label may be a good thing, as early adopters of the rolling pre-freeze
+releases are likely to encounter unexpected consequences at the level of the
+wider Python ecosystem, and would need to take an active part in getting those
+issues resolved.
```
Maybe we could reframe this as "our expectation is that it will take time to build trust in the new labels, and so slower adoption of the pre-freeze releases due to lingering fears around the names is a good thing" rather than it being treated as "OMG nobody will use it because it says 'beta'"?

ncoghlan replied:
Exactly - the *first* time we do this, we want to limit the audience to folks that are willing to be active participants in making it work in the real world.

If that proves feasible, then the second and subsequent times are when it makes sense to start removing incidental barriers to adoption (while still making it clear it isn't going to be suitable for everyone)

zooba commented on pep-0605.rst:
```diff
@@ -618,20 +710,31 @@ provisional CPython API instead.
+supporting the stable releases.
+
+For publishers of pure Python packages, this would be a matter of publishing
+unversioned wheel archives, and potentially adding the rolling pre-freeze
```
```suggestion
"py3" tagged wheel archives, and potentially adding the rolling pre-freeze
```

ncoghlan commented:
@zooba I've accepted your suggested changes, but your other comments will require more thought and editing. If you're happy to push the current version of the PR as an interim update until, please go ahead and merge.

ncoghlan commented on pep-0605.rst:
```diff
@@ -24,32 +24,43 @@ pre-release management process as follows:
-  releases are made, it is expected that they would be in place of one of the
-  scheduled beta releases
+* Alpha releases would be "ABI breaking pre-releases", where extension modules
+  compiled against the full CPython API in an earlier pre-release may no
```
Reworded

ncoghlan commented:
OK, I actually addressed the more substantive comments as well. It's a holiday weekend over here, so please merge if you're happy with the updates.

ncoghlan commented:
Antoine raised a valid point regarding duplication between the Abstract and the Proposal, so I cut the detailed list of bullet points out of the Abstract.

ncoghlan commented:
Added some images based on @ambv's example timeline and support matrix impact in PEP 602.

I also moved the example timeline up the top to reduce the initial "wall of text" impression.

ambv commented:
Good job on the pictures, they are helpful. One questionz: shouldn't 3.10 start in June rather than May? E.g. branch cut happens at RC?

ncoghlan commented:
@ambv Thanks for providing the base images for me to start from! (I'm completely hopeless when I'm looking at a blank screen in an image editor).

Under the current text, there's a 4 month preparation window on master between branching the previous RC, and publishing the initial ABI for the next release series. I'll see what I can do to make it clear that that's intentional (the "2 month RC preparation" window actually means skipping two pre-releases, one at the start, and one at the end, but it isn't obvious that that results in having 4 months to prepare the first alpha for the next series)

ncoghlan commented:
I went ahead and merged the update, as the revisions here were substantial enough that reading the previously published version wouldn't give a good understanding of the current state of the proposal.
